ballerina/memory.bal

```diff
@@ -75,18 +155,60 @@ public isolated class MessageWindowChatMemory {
         lock {
             self.createSessionIfNotExist(sessionId);
-            MemoryChatMessage[] memory = self.sessions.get(sessionId);
-            if memory.length() >= self.size - 1 {
-                _ = memory.shift();
-            }
             if newMessage is MemoryChatSystemMessage {
                 self.systemMessageSessions[sessionId] = newMessage;
                 return;
             }
+
+            MemoryChatMessage[] memory = self.sessions.get(sessionId);
+            int memoryLength = memory.length();
+            ROLE newMessageRole = newMessage.role;
+            int size = self.size;
+            SummarizeOverflowConfig? config = self.summarizeOverflowConfig;
+
+            if memoryLength >= size - 1 {
+                if config is () || size <= 3 {
+                    // No summarization config or too small window - just remove oldest
+                    _ = memory.shift();
+                } else {
+                    ROLE lastMessageRole = memory[memoryLength - 1].role;
+                    boolean isLastInteractionFromAgent = lastMessageRole != USER && newMessageRole == USER;
+                    MemoryChatMessage[] slicedMemory = isLastInteractionFromAgent
+                        ? memory.slice(0, memoryLength - 1)
+                        : memory;
+
+                    MemoryChatMessage|Error summaryMessage = self.generateSummary(
+                        slicedMemory, config.modelProvider, config.summarizationPrompt, config.maxSummaryTokens);
+                    if summaryMessage is Error {
+                        // Fallback: remove oldest message if summarization fails
+                        _ = memory.shift();
+                    } else {
+                        if isLastInteractionFromAgent {
+                            MemoryChatMessage lastMessage = memory.pop();
+                            memory.removeAll();
+                            memory.push(lastMessage);
+                        } else {
+                            memory.removeAll();
+                        }
+                        memory.unshift(summaryMessage);
+                        self.sessions[sessionId] = memory;
+                    }
+                }
+            }
+
             memory.push(newMessage);
         }
     }
 
+    isolated function generateSummary(MemoryChatMessage[] slicedMemory, ModelProvider provider,
+            Prompt summarizationPrompt, int maxSummaryTokens) returns MemoryChatMessage|Error {
+        string updatedPrompt = CHAT_HISTORY_REGEX.replace(stringifyPromptContent(
+                summarizationPrompt), slicedMemory.toString());
+        updatedPrompt = MAX_SUMMARY_TOKEN_COUNT_REGEX.replace(updatedPrompt, maxSummaryTokens.toString());
+        ChatAssistantMessage summarizationModelResult = check callSummarizationModel(provider, updatedPrompt);
+        return {role: USER, content: string `${SUMMARY_PREFIX} ${summarizationModelResult.content.toString()}`};
```

Comment:
Wouldn't this produce invalid content if `summarizationModelResult.content` is nil?